### ISSUE TITLE
Asynchronous API

### ISSUE THREAD

**Issue by alenkacz** [rabbitmq/hop#103]
Any plans on making the library asynchronous? (returning futures instead of blocking and returning values directly) ... ?

**Comment by acogoluegnes** [comment]
No plan of doing this, but it's worth discussing about it. Asynchronous APIs usually help to scale (doing more with fewer threads). This library is mostly used for monitoring (as far as I know), so there's not too much pressure on the client side. What would be your use case?

**Comment by michaelklishin** [comment]
This library is intended for monitoring and in many ways it is efficient enough for that. RabbitMQ Java client has dropped JDK 6 support recently and is a `CompletableFuture`-based API is in the works (we will see to what extent it can be introduced to a 10 year old library).

This client is significantly smaller and younger, so doing the same here should be in theory be easier. As is developing a new HTTP API client for those who need a futures-based API ;)

**Comment by artembilan** [comment]
Well, we may even consider to add an API based on a new Spring's WebFlux Client: http://docs.spring.io/spring/docs/5.0.0.RC1/spring-framework-reference/web.html#web-reactive

**Comment by michaelklishin** [comment]
@artembilan good idea. So yeah, once Spring 5 ships, I think one way or another we'd like to add an asynchronous version and drop JDK 6 support (supporting 8+ only).

**Comment by michaelklishin** [comment]
`effort-high` because porting the entire API and all the tests will likely take a few weeks even with WebFlux Client.

**Comment by michaelklishin** [comment]
@alenkacz we have merged a reactive client in master. Please give it a shot and let us know how it goes :)
